### PR TITLE
#1756 better forged method error based on empty target bean properties

### PR DESCRIPTION
**Pull request by sjaakd**

### CODE REVIEW

**Comment by chris922** [comment]
What is now the difference? You just added "No target bean properties found" at the beginning of the message?

**Comment by sjaakd** [comment]
> What is now the difference? You just added "No target bean properties found" at the beginning of the message?

exactly.. I was puzzled when I got this warning. I had 2 beans defined (empty albeit), but I had to go deep into tracing to figure out what the reason was.

**Comment by sjaakd** [comment]
@chris922 : thanks for reviewing 👍

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 1]
```diff
@@ -39,7 +39,8 @@
     PROPERTYMAPPING_CREATE_NOTE( "creating property mapping: %s.", Diagnostic.Kind.NOTE ),
     PROPERTYMAPPING_SELECT_NOTE( "selecting property mapping: %s.", Diagnostic.Kind.NOTE ),
     PROPERTYMAPPING_MAPPING_NOT_FOUND( "Can't map %s to \"%s %s\". Consider to declare/implement a mapping method: \"%s map(%s value)\"." ),
-    PROPERTYMAPPING_FORGED_MAPPING_NOT_FOUND( "Can't map %s to %s. Consider to implement a mapping method: \"%s map(%s value)\"." ),
+    PROPERTYMAPPING_FORGED_MAPPING_WITH_HISTORY_NOT_FOUND( "No target bean properties found: can't map %s to \"%s %s\". Consider to declare/implement a mapping method: \"%s map(%s value)\"." ),
+    PROPERTYMAPPING_FORGED_MAPPING_NOT_FOUND( "No target bean properties found: can't map %s to %s. Consider to implement a mapping method: \"%s map(%s value)\"." ),
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: This change is not OK. The `PROPERTYMAPPING_FORGED_MAPPING_NOT_FOUND` is also used in other places where this error message is not valid.

File: processor/src/test/java/org/mapstruct/ap/test/collection/forged/CollectionMappingTest.java [hunk 2]
```diff
@@ -87,8 +87,8 @@ public void shouldForgeNewMapMappingMethod() {
             @Diagnostic(type = ErroneousCollectionNonMappableSetMapper.class,
                 kind = Kind.ERROR,
                 line = 17,
-                messageRegExp = "Can't map Collection element \".* nonMappableSet\" to \".* nonMappableSet\". "
-                    + "Consider to declare/implement a mapping method: .*."),
+                messageRegExp = "No target bean properties found: can't map Collection element \".* nonMappableSet\" "
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/collection/forged/CollectionMappingTest.java]: This should not have been changed. See my previous comment

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 1 above)
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: I'll be happy to open it up again.. But I can only find one reference of both messages.. Its in

```java
            if ( method instanceof ForgedMethod && targetProperties.isEmpty() ) {
                //TODO until we solve 1140 we report this error when the target properties are empty
                ForgedMethod forgedMethod = (ForgedMethod) method;
                if ( forgedMethod.getHistory() == null ) {
                    Type sourceType = this.method.getParameters().get( 0 ).getType();
                    Type targetType = this.method.getReturnType();
                    ctx.getMessager().printMessage(
                        this.method.getExecutable(),
                        Message.PROPERTYMAPPING_FORGED_MAPPING_NOT_FOUND,
                        sourceType,
                        targetType,
                        targetType,
                        sourceType
                    );
                }
                else {
                    ForgedMethodHistory history = forgedMethod.getHistory();
                    ctx.getMessager().printMessage(
                        this.method.getExecutable(),
                        Message.PROPERTYMAPPING_FORGED_MAPPING_WITH_HISTORY_NOT_FOUND,
                        history.createSourcePropertyErrorMessage(),
                        history.getTargetType(),
                        history.createTargetPropertyName(),
                        history.getTargetType(),
                        history.getSourceType()
                    );
                }
            }
```

Look at the line ` if ( method instanceof ForgedMethod && targetProperties.isEmpty() ) `.. that coincides with the new message, right?

File: processor/src/test/java/org/mapstruct/ap/test/collection/forged/CollectionMappingTest.java (hunk 2 above)
Comment (sjaakd) [reply processor/src/test/java/org/mapstruct/ap/test/collection/forged/CollectionMappingTest.java]: I'm not so sure.. the full error message (removing the wild card matches is:

`No target bean properties found: can't map Collection element "org.mapstruct.ap.test.collection.forged.Foo nonMappableSet" to "org.mapstruct.ap.test.collection.forged.Bar nonMappableSet". Consider to declare/implement a mapping method: "org.mapstruct.ap.test.collection.forged.Bar map(org.mapstruct.ap.test.collection.forged.Foo value)`

`Bar` is an empty bean (No properties).

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 1 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Ah yes you are right. I searched for the wrong message. I was searching for `PROPERTYMAPPING_MAPPING_NOT_FOUND` and not for `PROPERTYMAPPING_FORGED_MAPPING_NOT_FOUND`. Sorry for the mistake

File: processor/src/test/java/org/mapstruct/ap/test/collection/forged/CollectionMappingTest.java (hunk 2 above)
Comment (filiphr) [reply processor/src/test/java/org/mapstruct/ap/test/collection/forged/CollectionMappingTest.java]: You are right, sorry for my mistake

**Pull request merged**